Add Side-Input Support in Flink Streaming Runner

aljoscha opened the pull request:
This does several things in somewhat isolated steps:
 - Unify the operator wrappers. Before there where 4 different DoFn/Window operator wrappers with code duplication. I unified this to make the implementation of side inputs for both `DoFn` and `CombineFnWithContext` easier.
 - Change `StateInternals` to use a Flink `AbstractStateBackend` internally. This is essentially the Flink version of `StateInternals` and moving on top of this will allow us to benefit from parallelism rescaling in the future. Also, it reduces the amount of code we have in the runner because we did all state serialization basically by hand before.
 - Add a generic `SideInputHandler` that uses `StateInternals` to keep the side input data as well as meta data. This is in the runners-core package and should be reusable for other runner implementations. The implementation is rather simple. The complicated part was all the plumbing and translation around it.
 - Change `ParDoTest.TestDoFnWithContext` to allow `DoFn` reuse. I think this might have been overlooked in the recent change to allow `DoFn` reuse. @tgroh ?
 - Change `PAssert` checker `DoFns` to not suppress exceptions in streaming mode. Otherwise the tests will not fail for the Flink runner when expected. Not sure if this is possible because the Dataflow runner needs it like that.

The rest of the changes are aimed at making the `RunnableOnService` tests execute without errors on the streaming runner, all of the commits contain a lot of small fixes towards that goal.

There is still the issue of side-input data not being garbage collected but I think this is a general problem and not specific to the Flink runner.

There are still two failing tests but I think the problem might actually be with the tests 😉:
 - `CombineTest.testGlobalCombineWithDefaultsAndTriggers` fails with
```
Expected: a collection containing "2: true"
     but: was "1: true"
```
I think this test works for the `DirectRunner` because the two input elements are all processed in one bundle and the Trigger is only asked after the bundle was processed whether we should fire. This produces the expected result. If you have a runner that asks the trigger after every element whether it should fire this means that the trigger fires after seeing the first element. Afterwards it goes into finished state and the pane is dropped and we never emit a pane at GC time as I would expect with ACCUMULATING mode. The relevant code is in `ReduceFnRunner.emitIfAppropriate`. `isFinished` will be true for the count trigger which in turn means `shouldDiscard` will be true.
 - `CombineTest.testHotKeyCombiningWithAccumulationMode` this test fails because it only expects one pane do be emitted while a tuple-at-a-time runner such as the Flink streaming runner will emit a pane for every element. For the `DirectRunner` it again works because all the elements are processed in one bundle and the trigger only fires once.

R: @mxm 
R: @tgroh for the change to the `ParDoTest` test and also overall
R: @kennknowles For the `SideInputHandler` and also overall, turns out this piece was the easiest to do in all of this ...

tgroh commented:
So, I haven't looked beyond the CombineTest note, but you're definitely right. The trigger in `testGlobalCombineDefaultsWithTriggers` should be `Repeatedly.forever(AfterPane.elementCountAtLeast(1))`, and the `testHotKeyCombiningWithAccumulationMode` should have a more forgiving assertion.

aljoscha commented:
By the way, this is still missing tests for the `SideInputReader`, `DoFnOperator` and `WindowDoFnOperator`. We shouldn't merge like this, I'll add those shortly.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java:
```diff
@@ -135,9 +167,59 @@ public FlinkPipelineOptions getPipelineOptions() {
+//      } catch (CannotProvideCoderException e) {
+//        throw new RuntimeException("CANNO PRO", e);
+//      }
+//      return outputT;
```
Can this be removed?

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java:
```diff
@@ -176,4 +259,368 @@ public String toString() {
+      try {
+        inputCoder.getKeyCoder().verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+//        runner.recordViewUsesNonDeterministicKeyCoder(this);
```
empty catch

kennknowles replied:
If I understand the flow of code & data correctly for this and the multimap case, the `KV<K, V>` are concatenated into a list which is then converted by `MapPCollectionView` (via `MapViewFn`) into a conventional Java `Map<K, V>` using Java-level equality.

So for now I don't think a deterministic coder is necessary for your runner. This has to change for scalable cross-language map side inputs, but it seem OK for now. So instead of calling `verifyDeterministic` and throwing away the exception, these 5 lines can just be wiped out (and below).

aljoscha replied:
Yep, I'll remove in a follow-up PR.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java:
```diff
@@ -176,4 +259,368 @@ public String toString() {
+      try {
+        inputCoder.getKeyCoder().verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+//        runner.recordViewUsesNonDeterministicKeyCoder(this);
```
empty catch

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java:
```diff
@@ -99,79 +105,51 @@
+    TRANSLATORS.put(Reshuffle.class, new ReshuffleTranslatorStreaming());
+    TRANSLATORS.put(GroupByKey.class, new GroupByKeyTranslator());
+//    TRANSLATORS.put(
+//        FlinkRunner.StreamingCombineWithoutSideInput.class, new CombinePerKeyTranslator());
```
Why is this commented out?

mxm commented:
Hi @aljoscha! Thanks for the PR. Looks very promising. I haven't had time to check it out in detail but I want to give it a thorough look next week!

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/sdk/util/SideInputHandler.java:
```diff
@@ -0,0 +1,238 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
```
I wanted to say that the use of this namespace in the runners-core module is just so that the move from SDK to runners-core is a separate change from the package rename. I think here it is best to use `org.apache.beam.runners.core` directly.

Eventually it should all be in this namespace, to avoid split packages. This helps, for example, with shading, and makes import enforcement rules work better, etc.

aljoscha commented:
Thanks for the initial review! The Combine transform is still commented out because the runner does not yet correctly determine whether a combine is possible. This depends on whether the window function can merge and on whether we have side inputs. I still have to fix that for this PR. Same goes for the non-deterministic-key-coder portions.

aljoscha commented:
I rebased this on the current master and also added tests for the new `DoFnOperator` and `SideInputHandler`. This should be good to merge now from my side.

I'm quite confident that the Flink side is good now, I had to change some code in other parts, though, that would need review from someone knowledgeable in Beam. These changes are in the commits after the two main implementation commits. I'm thinking @tgroh, @dhalperi or @kennknowles  for this. If one of you would have a look that would be perfect. (The one I'm most unsure about is this one: e35b77d3bb61fa8dc59c9bf28f62ae57e07eb45b).

Also, both `SideInputHandler` and `SideInputHandlerTest` would probably benefit from another pass over them.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java:
```diff
@@ -176,4 +225,331 @@ public String toString() {
+      try {
+        inputCoder.getKeyCoder().verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+//        runner.recordViewUsesNonDeterministicKeyCoder(this);
```
empty catch. How about throwing an exception here because we don't support non-deterministic coders?

kennknowles replied:
empty catch here, too

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java:
```diff
@@ -59,14 +59,25 @@ OutputT apply(PTransform<InputT,OutputT> transform, InputT input) {
+      if (cause instanceof AssertionError) {
+        throw (AssertionError) cause;
       } else {
         throw e;
```
What is the purpose of the AssertionError check? You throw the exception either way.

aljoscha replied:
Jip, that seems to be leftover from something.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java:
```diff
@@ -0,0 +1,516 @@
+  protected final Collection<PCollectionView<?>> sideInputs;
+  protected final Map<Integer, PCollectionView<?>> sideInputTagMapping;
+
+  protected final boolean hasSideInputs;
```
It looks like this field is assigned but never used anywhere.

kennknowles commented:
Just back from a little vacation. I'll take a further look.

mxm commented:
It's tricky to review some of the changed parts because a lot of the files have been Checkstyle adjusted and renamed which makes them appear as completely new. After a second look I'm convinced that you did some really nice work here @aljoscha. Not only does it simplify a lot of the code but it also introduces a good abstraction to realize side inputs on top of Flink streaming which does not support side inputs natively. So +1 for merging from my side.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java:
```diff
@@ -0,0 +1,240 @@
+  /** The list of side inputs that we're handling. */
+  protected final Collection<PCollectionView<?>> sideInputs;
+
+  /** State internals that are scoped not to the key of a value but instead to one key group. */
```
Another way of describing this is that side inputs are, rather by definition, global data. So there's no relevant key.

aljoscha replied:
Jip, that's how I see them and that's the reason why I changed to `StateInternals<Void>` at some point. The comment is a leftover from when it was still `StateInternals<Integer>` and I thought about it as per key-group.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java:
```diff
@@ -0,0 +1,240 @@
+      PCollectionView<?>,
+      StateTag<
+          Object,
+          AccumulatorCombiningState<
```
BagState?

kennknowles replied:
I guess you want a `Set`.

aljoscha replied:
So you're saying "keep it as is"?

kennknowles commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java:
```diff
@@ -347,27 +314,292 @@ public void translateNode(
+
+        context.setOutputDataStream(context.getOutput(transform), outDataStream);
+      } else {
+        Tuple2<Map<Integer, PCollectionView<?>>, DataStream<RawUnionValue>> transformedSideInputs =
```
`RawUnionValue` should probably be in `sdk.values` or actually moved to `runners-core` to be shared by this `CoGBK` implementation and the runner. It is just a trivial class, but today is mostly supposed to be internal to the `CoGBK` implementation so never should have been public (but I guess it has to for coder purposes or something).

kennknowles replied:
Doesn't have to be in this PR, just a note, since the whole CoGBK situation is going to change.

kennknowles commented:
LGTM from me too.

Normally I might wait for the touch-ups we've commented on, but since this PR is already huge and the comments are not critical, I think it is best to merge and roll forwards unless someone has an objection. I will go ahead and merge it after a short waiting period.

mxm commented:
Thanks for merging. I think it is always better to address comments before merging. Everything that is not addressed from my experience either never gets fixed or comes back in forms of bug reports which could be non-trivial to debug (e.g. empty catch block). That said, let me see if I can address the minor issues that we commented on :)

aljoscha commented:
Thanks for the review @mxm and @kennknowles. I would quickly propose a follow up PR that also fixes some javadocs based on your comments.
